### ISSUE TITLE
`dependencies` and `devDependencies` for the Javascript web frontend UI 

### ISSUE THREAD

**Issue by fabcor-maxiv** [mxcube/mxcube3#1114]
I guess I always misunderstood how `dependencies` and `devDependencies` of `package.json` should be handled in a *React* project...

How come the following fails? At least it fails for me.

```
cd ui/
rm --force --recursive node_modules/
pnpm install --prod
pnpm run build
```

Does it really matter in such a project (that has a *build* step) to have `devDependencies` separated from `dependencies`? Or should we list all dependencies under `dependencies` and leave `devDependencies` empty?

Should we not use the `--prod`flag of `pnpm install`?

@axelboc

**Comment by axelboc** [comment]
Tldr:

- don't use the `--prod` flag;
- don't delete `node_modules` (except if you're installing with `pnpm` for the first time);
- keep the distinction between `dependencies` and `devDependencies` in `package.json`

---

### `dependencies` vs `devDependencies`

It's true that the distinction between `dependencies` and `devDependencies` makes more sense for libraries than for apps; and yes, in theory, if you're installing a library as a dependency, then that library should not need to have its dev dependencies installed to work. In practice, however, there are exceptions: some libraries have `postinstall` scripts (or similar) that build stuff with their dev dependencies after being installed.

For apps, the benefit is more in the readability of `package.json`: it helps identify dependencies that are used at runtime and dependencies that are used at build time. When you see 50 runtime dependencies, like in MXCuBE-Web, you start to really worry about bundle size and long-term maintainability, for instance... :joy: ; The separation can also be useful when prioritising which dependencies to upgrade, as runtime dependencies are obviously more critical.

### `--prod` flag
With pnpm, trying to reduce the number of dependencies that get installed with `--prod` does not make much sense.

pnpm uses a global store, which means that each dependency is only ever downloaded and stored once (per version) on your system. When you run `pnpm install`, if a package is already present in the global store, all it does is create a symlink to it. You no longer have `node_modules` folders everywhere, each taking up multiple GB of space like you did with npm.

It's not worth trying to use `--prod` on the CI either, since the pnpm store is typically cached, which means that most installs only take a few seconds.

---

Given:

- the likelihood that a dependency somewhere in the project's dependency graph needs its dev dependencies to run a `postinstall` script;
- **EDIT**: the likelihood of a package having a runtime dependency specified under `devDependencies` ... :joy: ;
- the readability benefit of separating runtime from build-time dependencies;
- how fast pnpm is;

I strongly recommend not using the `--prod` flag ever, and keeping `dependencies` from `devDependencies` separated in `package.json`.

**Comment by fabcor-maxiv** [comment]
@axelboc 

Thanks for the detailed answer, much appreciated! :)

1. For my understanding: if we were to put everything in `dependencies` or `devDependencies` the `pnpm run build` step would still produce the same result, is that correct?

1. How do we know whether a dependency is a runtime dependency (that should be listed in `dependencies`) or not (lint, format, test, build, and so on that should be listed in `devDependencies`). I mean of course we know, but how do we really know for sure? Is there some kind of linter rule that we can activate to tell us that we have a dependency in the wrong list?

1. We do not typically remove the `node_modules` directory, it was only meant to show how to reproduce the issue. Typically our CI pipeline starts without this directory. I guess we should look into configuring some cache for these things more seriously. The speed of `pnpm install` itself is not a priority for us right now, but it would still be better to have this setup more correctly. It is a task for another day I guess, and something we should solve internally at MAX IV.

**Comment by fabcor-maxiv** [comment]
I guess this is what I wanted to hear:

> I don't think npm's advice is very relevant here. It is primarily concerning Node apps. CRA doesn't give you a Node app. From that perspective, all dependencies (including React) are "dev" dependencies because they're only necessary for the build: once you build the app, it has no deps at all. But putting react and react-dom in devDependencies will be just as confusing, and create those additional problems for some hosting providers. So instead we put them all in dependencies.

-- https://github.com/facebook/create-react-app/issues/4342#issuecomment-383432122

> Again, unless you’re building a Node.js app, the most correct semantics would be to put everything into devDependencies including react and any libraries you’re using. But there are practical reasons why this is undesirable so we just went with the opposite approach.

-- https://github.com/facebook/create-react-app/issues/2696#issuecomment-320538668

**Comment by axelboc** [comment]
1. You're indeed correct, no impact at all.
2. Since it's more of a convention than anything else, and since there's no easy way to check for correctness programmatically, you have to decide yourself on a case-by-case basis. It can be difficult sometimes, for instance, you could argue that `react-scripts`—since it uses `webpack` internally, which injects runtime code for module loading—could go into `dependencies`... of course you wouldn't, it's a contrived example, but you get me. My point is: it has no real impact; it's just informative; so best effort from developers is good enough.
3. I've already put [caching of the pnpm store](https://github.com/mxcube/mxcubeweb/blob/d277c50f639d03c23a55cedae7658bc4e98a6c1e/.github/workflows/ui.yml#L36-L43) on the CI, so all is well on that front :blush:

**Comment by fabcor-maxiv** [comment]
@axelboc

Thanks. That answers my questions. Case closed, I guess :)

For the third point, I mean that we have our own internal CI/CD pipelines at MAX IV where there is no such caching configured yet. When we get to it we will surely take inspiration from how you solved this in the GitHub Actions workflow.
